clientd3d/d3drender_objects.c

```diff
@@ -1222,10 +1224,32 @@ void D3DRenderObjectsDraw(d3d_render_pool_new* pPool, room_type* room,
 	// count of objects found at that location.
 	std::unordered_map<int64, int> depth_adjustment_map;
 
+	// As we receive objects in different orders from the BSP walk this can cause inconsistent z-depth ordering.
+	// We now attempt to maintain a consistent view by sorting draw data by their ids.
+	std::sort(drawdata, drawdata + nitems, [](const DrawItem& a, const DrawItem& b) {
+
+		if (a.type != DrawObjectType && b.type == DrawObjectType) {
+			return false;
+		}
+		if (a.type == DrawObjectType && b.type != DrawObjectType) {
+			return true;
+		}
+
+		// If both items are of the DrawObjectType, compare their draw.id
+		if (a.type == DrawObjectType && b.type == DrawObjectType) {
+			// Compare draw.id from the object union member
+			return a.u.object.object->draw.id < b.u.object.object->draw.id;
+		}
+
+		// If the types are the same and not DrawObjectType, keep the original order.
+		// note: drawdata[] may contain more than one entry with the same id so this is not a stable sort.
```

Comment:
This comment is in the wrong place.  The sort isn't stable because of the comparison on line 1241.  If there are two objects with the same draw ID, they are ordered nondeterministically.